csrc/host_ir/host_ir.cpp

```diff
@@ -454,11 +462,57 @@ std::string ShardByStream::toInlineString(int indent_size) const {
 TensorView* shardByStream(TensorView* in, Val* stream_index) {
   auto* out = ops::newValLike(in, *in->getDataType())->as<TensorView>();
 
+  NVF_ERROR(
+      getShardedIterDomain(in, ParallelType::Stream) == nullptr,
+      "Input allocation shouldn't be sharded on stream: ",
+      in);
   TransformReplay::selfReplay(in->domain(), out->domain());
-  // This is conservative and suboptimal. Consider reusing the algorithm in
-  // https://github.com/NVIDIA/Fuser/blob/33337e9b0b82dc88bc305d9956101f0c8a8a0c60/csrc/alias_analysis.cpp#L199
-  // to decide contiguity.
-  out->setAllocationDomain(out->getLoopDomain(), false);
+
+  shardAllocationAsLoop(out, {ParallelType::Stream});
+  NVF_ERROR(
+      getShardedIterDomain(out, ParallelType::Stream) != nullptr,
+      "Output allocation should be sharded on stream after "
+      "shardAllocationAsLoop: ",
+      out);
+
+  // Refine the contiguity flags so `out` aliases `in`. This is done similar to
```

Comment:
Should this be in `shardAllocationAsLoop`, since we set the contiguity there as well.